pkg/core/function_test.go

```diff
@@ -14,106 +14,185 @@
-	const sleepsBeforeTimeout = 50
+var _ = Describe("Function", func() {
 
 	var (
-		errChan       = make(chan error, 1)
-		timeout       = sleepsBeforeTimeout * time.Microsecond
-		sleepDuration = time.Microsecond
-		log           io.Writer
-		err           error
-
-		hardErr                         = errors.New("hard error")
-		transientErr                    = errors.New("transient error")
-		transientErrCount, hardErrCount int
-		transientErrors, hardErrors     int
-
-		check = func() (error, error) {
-			transientErrCount++
-			if transientErrCount <= transientErrors {
-				return transientErr, nil
-			}
-			hardErrCount++
-			if hardErrCount <= hardErrors {
-				return nil, hardErr
-			}
-			return nil, nil
-		}
+		client               core.Client
+		mockClientConfig     *vendor_mocks.ClientConfig
+		mockBuildFactory     *mockbuildfactory.BuildFactory
+		mockBuild            *mockbuildfactory.Build
+		mockServing          *mockserving.Interface
+		mockServingV1alpha1  *mockserving.ServingV1alpha1Interface
+		mockServiceInterface *mockserving.ServiceInterface
+		workDir              string
+		service              *v1alpha1.Service
+		testService          *v1alpha1.Service
+		err                  error
 	)
 
 	BeforeEach(func() {
-		log = ioutil.Discard
-		transientErrCount = 0
-		hardErrCount = 0
-		transientErrors = 0
-		hardErrors = 0
+		mockClientConfig = &vendor_mocks.ClientConfig{}
+		mockBuildFactory = &mockbuildfactory.BuildFactory{}
+		mockBuild = &mockbuildfactory.Build{}
+		mockServing = &mockserving.Interface{}
+		mockServingV1alpha1 = &mockserving.ServingV1alpha1Interface{}
+		mockServiceInterface = &mockserving.ServiceInterface{}
+		mockServing.On("ServingV1alpha1").Return(mockServingV1alpha1)
+		mockServingV1alpha1.On("Services", mock.Anything).Return(mockServiceInterface)
+		testService = &v1alpha1.Service{}
+		workDir = test_support.CreateTempDir()
+		mockClientConfig.On("Namespace").Return("default", false, nil)
+		client = core.NewClient(mockClientConfig, nil, nil, mockServing, mockBuildFactory)
 	})
 
-	JustBeforeEach(func() {
-		err = pollService(check, errChan, timeout, sleepDuration, log)
+	AfterEach(func() {
+		test_support.CleanupDirs(GinkgoT(), workDir)
 	})
 
-	Context("when an error is sent to the error channel", func() {
-		BeforeEach(func() {
-			errChan <- errors.New("channel error")
-		})
-
-		It("should stop polling and return the error", func() {
-			Expect(err).To(MatchError("channel error"))
-		})
-	})
+	Describe("CreateFunction", func() {
+		var (
+			createFunctionOptions core.CreateFunctionOptions
+			createdService        *v1alpha1.Service
+		)
 
-	Context("when a hard error occurs", func() {
 		BeforeEach(func() {
-			hardErrors = 1
+			mockServiceInterface.On("Create", mock.Anything).Return(testService, nil).Run(func(args mock.Arguments) {
+				createdService = args.Get(0).(*v1alpha1.Service)
+			}) // input to Create() should be thing returned by function create
```

Comment:
Not sure I understand this. Stores the arg in `createdService`, but returns `testService`
Also, may be more readable to swap the calls to Return and Run

Reply:
> Not sure I understand this. Stores the arg in createdService, but returns testService
I'll improve the comment.